# Data Cleaning

## Documenting the Assessment

1. Column names should not contain the space. (Remove or replace spaces from the column names).
2. Incorrect Data types:
    1. `Order Date` and `Ship Date` columns have str dtype (Change it to datetime)
    2. `Postal Code` have int64 dtype (Change it to str)
    3. `Ship Mode`, `Segment`, `Region`, `Category`, `State` & `Sub-Category` have dtype str (Change it to Categorical)
3. Order ID `US-2014-150119` has one duplicate record. 

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('../data/raw-data/superstore.csv', encoding='latin1')

In [3]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


# Cleaning

### Lets fix nulls

In [4]:
# Check
df.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

### Fix Column names

In [7]:
# Column names and clean column names dic
rename_dict = {
    'Row ID': 'row-id',
    'Order ID': 'order-id',
    'Order Date': 'order-date',
    'Ship Date': 'ship-date',
    'Ship Mode': 'ship-mode',
    'Customer ID': 'customer-id',
    'Customer Name': 'customer-name',
    'Segment': 'segment',
    'Country': 'country',
    'City': 'city',
    'State': 'state',
    'Postal Code': 'postal-code',
    'Region': 'region',
    'Product ID': 'product-id',
    'Category': 'category',
    'Sub-Category': 'sub-category',
    'Product Name': 'product-name',
    'Sales': 'sales',
    'Quantity': 'quantity',
    'Discount': 'discount',
    'Profit': 'profit'
}

In [8]:
df.rename(columns=rename_dict, inplace=True)

In [9]:
# Check
df.head()

,row-id,order-id,order-date,ship-date,ship-mode,customer-id,customer-name,segment,country,city,...,postal-code,region,product-id,category,sub-category,product-name,sales,quantity,discount,profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


### Fix column data types

#### Fixing `order-date` & `ship-date`

In [10]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   row-id         9994 non-null   int64  
 1   order-id       9994 non-null   str    
 2   order-date     9994 non-null   str    
 3   ship-date      9994 non-null   str    
 4   ship-mode      9994 non-null   str    
 5   customer-id    9994 non-null   str    
 6   customer-name  9994 non-null   str    
 7   segment        9994 non-null   str    
 8   country        9994 non-null   str    
 9   city           9994 non-null   str    
 10  state          9994 non-null   str    
 11  postal-code    9994 non-null   int64  
 12  region         9994 non-null   str    
 13  product-id     9994 non-null   str    
 14  category       9994 non-null   str    
 15  sub-category   9994 non-null   str    
 16  product-name   9994 non-null   str    
 17  sales          9994 non-null   float64
 18  quantity       9994

In [12]:
df['order-date'] = pd.to_datetime(df['order-date'], format='%m/%d/%Y')
df['ship-date'] = pd.to_datetime(df['ship-date'], format='%m/%d/%Y')

In [13]:
#Check
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   row-id         9994 non-null   int64         
 1   order-id       9994 non-null   str           
 2   order-date     9994 non-null   datetime64[us]
 3   ship-date      9994 non-null   datetime64[us]
 4   ship-mode      9994 non-null   str           
 5   customer-id    9994 non-null   str           
 6   customer-name  9994 non-null   str           
 7   segment        9994 non-null   str           
 8   country        9994 non-null   str           
 9   city           9994 non-null   str           
 10  state          9994 non-null   str           
 11  postal-code    9994 non-null   int64         
 12  region         9994 non-null   str           
 13  product-id     9994 non-null   str           
 14  category       9994 non-null   str           
 15  sub-category   9994 non-null   s

#### Fixing `postal-code`

In [14]:
# comvert into str
df['postal-code'] = df['postal-code'].astype('str')

In [16]:
df['postal-code'].sample(5)

8168    98105
6495    77095
1859     6010
2649    74133
393     61604
Name: postal-code, dtype: str

In [17]:
# For postal code that are of length 4 we have to make it 5 by filling leading 0
df['postal-code'] = df['postal-code'].str.zfill(5)

In [35]:
df['postal-code'].sample(5)

9542    77095
4617    02149
7911    95037
1767    27405
579     60610
Name: postal-code, dtype: str

#### Fixing `Ship Mode`, `Segment`, `Region`, `Category`, `State` & `Sub-Category`.

In [36]:
df['ship-mode'] = df['ship-mode'].astype('category')
df['segment'] = df['segment'].astype('category')
df['state'] = df['state'].astype('category')
df['region'] = df['region'].astype('category')
df['category'] = df['category'].astype('category')
df['sub-category'] = df['sub-category'].astype('category')

In [37]:
# Check
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   row-id         9994 non-null   int64         
 1   order-id       9994 non-null   str           
 2   order-date     9994 non-null   datetime64[us]
 3   ship-date      9994 non-null   datetime64[us]
 4   ship-mode      9994 non-null   category      
 5   customer-id    9994 non-null   str           
 6   customer-name  9994 non-null   str           
 7   segment        9994 non-null   category      
 8   country        9994 non-null   str           
 9   city           9994 non-null   str           
 10  state          9994 non-null   category      
 11  postal-code    9994 non-null   str           
 12  region         9994 non-null   category      
 13  product-id     9994 non-null   str           
 14  category       9994 non-null   category      
 15  sub-category   9994 non-null   c

### Fixing duplicate data

In [42]:
df[df['order-id']=='US-2014-150119']

,row-id,order-id,order-date,ship-date,ship-mode,customer-id,customer-name,segment,country,city,...,postal-code,region,product-id,category,sub-category,product-name,sales,quantity,discount,profit
3405,3406,US-2014-150119,2014-04-23,2014-04-27,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,...,43229,East,FUR-CH-10002965,Furniture,Chairs,Global Leather Highback Executive Chair with P...,281.372,2,0.3,-12.0588
3406,3407,US-2014-150119,2014-04-23,2014-04-27,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,...,43229,East,FUR-CH-10002965,Furniture,Chairs,Global Leather Highback Executive Chair with P...,281.372,2,0.3,-12.0588
3407,3408,US-2014-150119,2014-04-23,2014-04-27,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,...,43229,East,OFF-BI-10000145,Office Supplies,Binders,Zipper Ring Binder Pockets,7.488,8,0.7,-5.2416
3408,3409,US-2014-150119,2014-04-23,2014-04-27,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,...,43229,East,FUR-FU-10002191,Furniture,Furnishings,G.E. Halogen Desk Lamp Bulbs,22.336,4,0.2,7.8176


We are not sure but it seems that the row-id 3406 and 3407 are exact same record but it is also possible that the customer ordered it two times. But as per my knowledge it is dplicate so I am removing the 2nd recode 3407  
It is ar 3406th index

In [47]:
df.drop(3406, inplace=True)

In [49]:
df.shape

(9993, 21)

In [50]:
df

,row-id,order-id,order-date,ship-date,ship-mode,customer-id,customer-name,segment,country,city,...,postal-code,region,product-id,category,sub-category,product-name,sales,quantity,discount,profit
0,1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9989,9990,CA-2014-110422,2014-01-21,2014-01-23,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,...,33180,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.2480,3,0.20,4.1028
9990,9991,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.9600,2,0.00,15.6332
9991,9992,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.5760,2,0.20,19.3932
9992,9993,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.6000,4,0.00,13.3200


### Drop `row-id` column
It is not usefull column for analysis so droping it

In [51]:
df.drop(columns='row-id', inplace=True)

In [52]:
df

,order-id,order-date,ship-date,ship-mode,customer-id,customer-name,segment,country,city,state,postal-code,region,product-id,category,sub-category,product-name,sales,quantity,discount,profit
0,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9989,CA-2014-110422,2014-01-21,2014-01-23,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,Florida,33180,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.2480,3,0.20,4.1028
9990,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,California,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.9600,2,0.00,15.6332
9991,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,California,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.5760,2,0.20,19.3932
9992,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,California,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.6000,4,0.00,13.3200


In [53]:
# Save the cleaned df into pkl and csv file
df.to_csv('../data/cleaned-data/cleaned.csv', index=False)
df.to_pickle('../data/cleaned-data/cleaned.pkl')